In [168]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

# load the dataset

In [225]:
df=pd.read_excel(r"C:\Users\disha\Downloads\Association Rules\Association Rules\Online retail.xlsx")
print(df)

     shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                             

# data preprocessing

# converting the data set into appropriate format

In [226]:
df.iloc[:, 0] = df.iloc[:, 0].astype(str)
transactions = df.iloc[:, 0].str.split(',')
data = pd.get_dummies(transactions.apply(pd.Series).stack()).groupby(level=0).sum()
print("One-Hot Encoded DataFrame:\n", data.head())


One-Hot Encoded DataFrame:
     asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0           0        0                  0          0        0            0   
1           0        0                  0          0        0            0   
2           0        0                  0          0        1            0   
3           0        0                  0          0        0            0   
4           0        0                  0          0        0            0   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0      0               0          0            0  ...       0               0   
1      0               0          0            0  ...       0               0   
2      0               0          0            0  ...       1               0   
3      0               0          0            0  ...       0               0   
4      0               0          0            0  ...       0               0   

   water spray  

In [227]:
data.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0           0        0                  0          0        0            0   
1           0        0                  0          0        0            0   
2           0        0                  0          0        1            0   
3           0        0                  0          0        0            0   
4           0        0                  0          0        0            0   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0      0               0          0            0  ...       0               0   
1      0               0          0            0  ...       0               0   
2      0               0          0            0  ...       1               0   
3      0               0          0            0  ...       0               0   
4      0               0          0            0  ...       0               0   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0            0           0                 0                  0   
1            0           0                 0                  0   
2            0           0                 0                  0   
3            0           0                 0                  0   
4            0           0                 0                  0   

   whole wheat rice  yams  yogurt cake  zucchini  
0                 0     0            0         0  
1                 0     0            0         0  
2                 0     0            0         0  
3                 1     0            0         0  
4                 0     0            0         0  

[5 rows x 120 columns]

In [236]:
data = data.astype(bool)
print("One-Hot Encoded DataFrame:\n", data.head())

One-Hot Encoded DataFrame:
     asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False    False              False      False    False        False   
1       False    False              False      False    False        False   
2       False    False              False      False     True        False   
3       False    False              False      False    False        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False           False   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...    True           False   
3  False           False      False        False  ...   False           False   
4  False           False      False        False  ...   False           False   

   water spray  

# Apriori Algorithm

In [245]:
frequent_itemsets = apriori(data, min_support=0.05, use_colnames=True)
print("Frequent Itemsets:\n", frequent_itemsets)

Frequent Itemsets:
      support                    itemsets
0   0.087200                   (burgers)
1   0.081067                      (cake)
2   0.060000                   (chicken)
3   0.163867                 (chocolate)
4   0.080400                   (cookies)
5   0.051067               (cooking oil)
6   0.179733                      (eggs)
7   0.079333                  (escalope)
8   0.170933              (french fries)
9   0.063200           (frozen smoothie)
10  0.095333         (frozen vegetables)
11  0.052400             (grated cheese)
12  0.132000                 (green tea)
13  0.098267               (ground beef)
14  0.076400            (low fat yogurt)
15  0.129600                      (milk)
16  0.238267             (mineral water)
17  0.065733                 (olive oil)
18  0.095067                  (pancakes)
19  0.071333                    (shrimp)
20  0.050533                      (soup)
21  0.174133                 (spaghetti)
22  0.068400                  (tomato

# association rule 

In [255]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.1)
print("Association Rules:\n", rules)

Association Rules:
        antecedents      consequents  antecedent support  consequent support  \
0      (chocolate)  (mineral water)            0.163867            0.238267   
1  (mineral water)      (chocolate)            0.238267            0.163867   
2           (eggs)  (mineral water)            0.179733            0.238267   
3  (mineral water)           (eggs)            0.238267            0.179733   
4      (spaghetti)  (mineral water)            0.174133            0.238267   
5  (mineral water)      (spaghetti)            0.238267            0.174133   

    support  confidence      lift  leverage  conviction  zhangs_metric  
0  0.052667    0.321400  1.348907  0.013623    1.122506       0.309351  
1  0.052667    0.221041  1.348907  0.013623    1.073398       0.339566  
2  0.050933    0.283383  1.189351  0.008109    1.062957       0.194090  
3  0.050933    0.213766  1.189351  0.008109    1.043286       0.209004  
4  0.059733    0.343032  1.439698  0.018243    1.159468      

In [257]:
rules.head()

antecedents      consequents  antecedent support  consequent support  \
0      (chocolate)  (mineral water)            0.163867            0.238267   
1  (mineral water)      (chocolate)            0.238267            0.163867   
2           (eggs)  (mineral water)            0.179733            0.238267   
3  (mineral water)           (eggs)            0.238267            0.179733   
4      (spaghetti)  (mineral water)            0.174133            0.238267   

    support  confidence      lift  leverage  conviction  zhangs_metric  
0  0.052667    0.321400  1.348907  0.013623    1.122506       0.309351  
1  0.052667    0.221041  1.348907  0.013623    1.073398       0.339566  
2  0.050933    0.283383  1.189351  0.008109    1.062957       0.194090  
3  0.050933    0.213766  1.189351  0.008109    1.043286       0.209004  
4  0.059733    0.343032  1.439698  0.018243    1.159468       0.369806

# Analysis and Interpretation:

Rule 1:
Support: 5.27% of all transactions contain both mineral water and chocolate.
Confidence: 22.10% of transactions that contain mineral water also contain chocolate.
Lift: 1.35 indicates that buying mineral water increases the likelihood of buying chocolate by 34.89%.
Leverage: 0.0136 shows the increase in the probability of chocolate purchases when mineral water is bought.
Conviction: 1.07 indicates a weak dependency.
Zhang's Metric: 0.34 shows a positive correlation.


Insight: Consider bundling these items together to boost sales. Placing them near each other in the store might increase the likelihood of combined purchases.


These insights help optimize sales strategies, enhance customer satisfaction, and increase overall sales by aligning with customer behavior patterns.